ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,933 @@
+/// the "left proof" also contains non-zero masks, `mask_p` and `mask_q`.
+/// For the "right proof" both masks are zero.
+///
+/// Proofs that are sent to other parties have mask that are always zero and therefore don't need to be sent.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct NIDZKP<F>
+where
+    F: Field,
+{
+    pub(crate) proofs: Vec<Vec<F>>,
+    pub(crate) mask_p: F,
+    pub(crate) mask_q: F,
+}
+
+/// generates recursive proof for `sum_i u_i*v_i = out` over field `F`
+///
+/// `r_f`, i.e. recursion factor, determines how much the inputs are compressed during each recursion
+/// it is also the degree of polynomials p and q
+///
+/// the output on the right needs to be sent to the party on the left
+/// it is therefore the "right share" of the party on the left
+///
+/// the left part of the output is the "left share" of the proof generated by the party on the right
+/// this part is generated using PRSS and does not need to be sent,
+/// it remains with the party and is the input `proof_left` to the proof verification
+/// the outputs includes the two masks for the proofs, i.e. `mask_p`, `mask_q`
+///
+/// This function uses `ctx` for generating random values using PRSS
+/// and sending the right part of the proof to the left
+///
+/// # Errors
+/// propagates errors from send and receive
+///
+/// # Panics
+/// Does not panic
+pub async fn generate_proof<C, F>(
+    ctx: C,
+    u: &mut Vec<F>,
+    v: &mut Vec<F>,
+    r_f: usize,
+) -> Result<(NIDZKP<F>, NIDZKP<F>), Error>
+where
+    C: Context,
+    F: Field + Invert,
+{
+    // check length
+    debug_assert_eq!(u.len(), v.len());
+    // check that field is large enough to hold evaluation points
+    debug_assert!(F::BITS > usize::BITS - r_f.leading_zeros() + 1);
+    // record, used as counter for PRSS
+    let mut record = 0usize;
+
+    // proof length
+    #[allow(
+        clippy::cast_possible_truncation,
+        clippy::cast_sign_loss,
+        clippy::cast_precision_loss
+    )]
+    let proof_length = (u.len() as f64).log(r_f as f64).ceil() as usize + 1usize;
+    // get masks from PRSS
+    let mask_p = ctx.prss().generate_fields(RecordId::from(record)); //(F::ONE, F::ONE);
+    record += 1;
+    let mask_q = ctx.prss().generate_fields(RecordId::from(record));
+    // proof from the prover on the left, i.e. generated via `PRSS.left` rather than receiving it from the left
+    let mut output_left = NIDZKP {
+        proofs: Vec::with_capacity(proof_length),
+        // set mask p(0) from `PRSS.left`
+        mask_p: mask_p.0,
+        // set mask q(0) from `PRSS.left`
+        mask_q: mask_q.0,
+    };
+    // right part of this helpers own proof, to be sent to the left
+    let mut own_proof_right = NIDZKP {
+        proofs: Vec::with_capacity(proof_length),
+        mask_p: F::ZERO,
+        mask_q: F::ZERO,
+    };
+
+    // precomputation for Lagrange
+    let mut e_p = vec![F::ZERO; 2 * r_f];
```

Comment:
I don't think you actually need to double it.

Let's take the example of `r_f = 4`, like I did in this Google Sheet: https://docs.google.com/spreadsheets/d/10tCSgiZyQbt4bMB3KqhAUOCfgfO8xIni83J2saipFCg/edit?usp=sharing, see the tab called "All Lagrange Interpolation".

In that case, the length of both `u` and `v` is 4. In all but the final round of recursion, we do not generate any random weights, meaning that each element of `u` and `v` can be interpolated with a degree 3 polynomial (maximum term is x^3).

So the Polynomial that is defined as `u /cdot v` is a degree 6 polynomial (maximum term is x^6). The uniquely define a degree 6 polynomial requires 7 points. We can easily compute 4 (by simply multiplying the 4 values of `u` and `v` provided), so we only need to compute 3 more. 

This is described in the 909 paper. In section 4.1 it says:

```
Next P_i computes a degree-2(λ − 1) polynomial...
```

I've worked it all out in the sheet. This is a nice optimization that saves one interpolation per iteration.